File: sword_to_offer/08.cpp
#include <iostream>

using namespace std;


struct BinaryTreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

BinaryTreeNode* GetNext(BinaryTreeNode* pNode){
    if(pNode = nullptr)
        return nullptr;

    BinaryTreeNode* pNext = nullptr;
    if(pNode->right != nullprt){
        BinaryTreeNode* pRight = pNode->right;
        while(pRight->left != left)
            pRight = pRight->left;
        pNext = pRight;
    }
    else if(pNode->parent != nullptr){
        BinaryTreeNode* pCurrent = pNode;
        BinaryTreeNode* pParent = pNode->parent;
        while(pParent != nullptr && pCurrent == pParent->right){
            pCurrent = pParent;
            pParent = pParent->parent;
        }
        pNext = pParent;
    }
    return pNext;
}

File: sword_to_offer/08.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def GetNext(node):
    if node == None:
        return None

    if node.right != None:
        pright = node.right
        while pright.left != None:
            pright = pright.left
        pnext = pright

    elif node.parent != None:
        pcurrent = node
        pparent = node.parent
        while pparent != None and pcurrent == pparent.right:
            pcurrent = pparent
            pparent = pparent.parent
        pnext = pparent

    return pnext
